include/cetl/pmr/interface_ptr.hpp

```diff
@@ -0,0 +1,92 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef CETL_PMR_INTERFACE_PTR_H_INCLUDED
+#define CETL_PMR_INTERFACE_PTR_H_INCLUDED
+
+#ifndef CETL_H_ERASE
+#    include "cetl/cetl.hpp"
+#endif
+
+#include "cetl/unbounded_variant.hpp"
+
+#include <memory>
+#include <functional>
+
+namespace cetl
+{
+namespace pmr
+{
+
+template <typename Interface>
+class PmrInterfaceDeleter final
+{
+public:
+    template <typename PmrAllocator>
+    PmrInterfaceDeleter(PmrAllocator alloc, std::size_t obj_count)
+    {
+        deleter_ = [alloc, obj_count](Interface* ptr) mutable {
+            using Concrete     = typename PmrAllocator::value_type;
+            auto* concrete_ptr = static_cast<Concrete*>(ptr);
+
+            concrete_ptr->~Concrete();
+            alloc.deallocate(concrete_ptr, obj_count);
+        };
+    }
+
+    template <typename Down>
+    PmrInterfaceDeleter(const PmrInterfaceDeleter<Down>& other)
```

Comment:
This is important to make possible up-casting of `InterfacePtr<B>` → `InterfacePtr<A` (assuming `class B : A {...}` inheritance).

Reply:
1. Upcasting as you described is unsafe to do via `static_cast` in the presence of data members. Why do we need to allow this conversion?

2. This definition hides the default copy and move constructors. I think it needs either sfinae or simply explicitly defaulted copy/move ctors (as they should take precedence over the template).

Reply:
> 1. Upcasting as you described is unsafe to...

Please correct me if I'm wrong, but I think this is actually DOWN-casting, and yes it's not safe generally. I need it to return back to the original pointer to concrete type (before calling its destructor and following deallocate at PMR). Without it following won't work:
```
class Concrete : Base /*sizeof f.e. 32*/, pubic ISomething
{
}
```
Instance of `InterfacePtr<ISomething>` (originally moved from `InterfacePtr<Concrete>`) stores internally pointer which is +32 bytes offset comparing to the original memory allocated (and initialized) for `Concrete` type instance.
So, when we are ready delete it then we need return back to the original pointer.